pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -534,26 +535,43 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, StablePoolStorage,
         uint256,
         uint256[] memory scalingFactors,
         bytes memory userData
-    ) internal override returns (uint256, uint256[] memory) {
+    ) internal override returns (uint256 bptAmountOut, uint256[] memory amountsIn) {
         StablePhantomPoolUserData.JoinKindPhantom kind = userData.joinKind();
 
         (uint256 virtualSupply, uint256[] memory balancesWithoutBpt) = _payProtocolFeesBeforeJoinExit(balances);
+        (uint256 currentAmp, ) = _getAmplificationParameter();
 
+        // The individual join functions mutate balancesWithoutBpt so that they reflect the post-join balances
         if (kind == StablePhantomPoolUserData.JoinKindPhantom.EXACT_TOKENS_IN_FOR_BPT_OUT) {
-            return _joinExactTokensInForBPTOut(virtualSupply, balancesWithoutBpt, scalingFactors, userData);
+            (bptAmountOut, amountsIn) = _joinExactTokensInForBPTOut(
+                virtualSupply,
+                currentAmp,
+                balancesWithoutBpt,
+                scalingFactors,
+                userData
+            );
         } else if (kind == StablePhantomPoolUserData.JoinKindPhantom.TOKEN_IN_FOR_EXACT_BPT_OUT) {
-            return _joinTokenInForExactBPTOut(virtualSupply, balancesWithoutBpt, userData);
+            (bptAmountOut, amountsIn) = _joinTokenInForExactBPTOut(
+                virtualSupply,
+                currentAmp,
+                balancesWithoutBpt,
+                userData
+            );
         } else {
             _revert(Errors.UNHANDLED_JOIN_KIND);
         }
+
+        // Pass in the post-join balances to reset the protocol fee basis.
+        _updateInvariantAfterJoinExit(currentAmp, balancesWithoutBpt);
```

Comment:
```suggestion
        // Add amountsIn to get post-join balances
        _mutateAmounts(balancesWithoutBpt, _dropBptItem(amountsIn), FixedPoint.add);
        
        // Pass in the post-join balances to reset the protocol fee basis.
        _updateInvariantAfterJoinExit(currentAmp, balancesWithoutBpt);
```

I would perform this mutation here rather than one layer down.

Reply:
Yes; much more self-contained